kura/org.eclipse.kura.asset.provider/src/main/java/org/eclipse/kura/asset/provider/BaseAsset.java

```diff
@@ -251,8 +265,13 @@ public AssetConfiguration getAssetConfiguration() {
         return this.assetConfiguration;
     }
 
-    public void setDriver(Driver driver) {
+    public synchronized void setDriver(Driver driver) {
         this.driver = driver;
+        if (driver != null) {
+            List<DriverRecord> readRecords = getAllReadRecords();
+            hasReadChannels = !readRecords.isEmpty();
+            tryPrepareRead(readRecords);
```

Comment:
tryPrepareRead() does not perform the actual read but just "prepares" it, it asks the driver to validate the current channel configuration and to perform eventual optimizations basing on it.
The idea behind prepared reads is to perform the operations above just once when the channel configuration changes and not on every read request.
To be able to do that we must have two things: the channel configuration and the driver instance attached.
This is the reason why tryPrepareRead() is called both in the updated() and in the setDriver() method: the first method provides the channel config and the second the driver instance, tryPrepareRead() checks whether we have both things and if yes it asks the driver to prepare the read.